From: Aleix Pol <aleixpol@kde.org>
Date: Fri, 06 May 2016 13:13:09 +0000
Subject: Only fetch the packages we actually need for PackageKit
X-Git-Url: http://quickgit.kde.org/?p=discover.git&a=commitdiff&h=8cc7eab119ea765442d6924915db419946e5cef4
---
Only fetch the packages we actually need for PackageKit

Should speed-up boot in some platforms where fetching is slower, such as
debian-based/apt.
---


--- a/libdiscover/backends/PackageKitBackend/PackageKitBackend.cpp
+++ b/libdiscover/backends/PackageKitBackend/PackageKitBackend.cpp
@@ -102,11 +102,15 @@
         disconnect(m_refresher.data(), &PackageKit::Transaction::finished, this, &PackageKitBackend::reloadPackageList);
     }
 
-    foreach(const Appstream::Component& component, m_appdata.allComponents()) {
+    const auto components = m_appdata.allComponents();
+    QStringList neededPackages;
+    neededPackages.reserve(components.size());
+    foreach(const Appstream::Component& component, components) {
         if (component.packageNames().isEmpty()) {
             qDebug() << "no packages for" << component.name();
             continue;
         }
+        neededPackages += component.packageNames();
 
         const auto res = new AppPackageKitResource(component, this);
         m_updatingPackages.packages[component.id()] = res;
@@ -118,8 +122,9 @@
             m_updatingPackages.extendedBy[pkg] += res;
         }
     }
-
-    PackageKit::Transaction * t = PackageKit::Daemon::getPackages();
+    neededPackages.removeDuplicates();
+
+    PackageKit::Transaction * t = PackageKit::Daemon::resolve(neededPackages);
     connect(t, &PackageKit::Transaction::finished, this, &PackageKitBackend::getPackagesFinished);
     connect(t, &PackageKit::Transaction::package, this, &PackageKitBackend::addPackage);
     connect(t, &PackageKit::Transaction::errorCode, this, &PackageKitBackend::transactionError);

